lib/api/apiUtils/object/objectLockHelpers.js

```diff
@@ -126,101 +130,175 @@ function setObjectLockInformation(headers, md, defaultRetention) {
+     */
+    isLocked() {
+        if (this.legalHold) {
             return true;
         }
-        const retentionMode = objectMD.retentionMode;
-        const retentionDate = objectMD.retentionDate;
-        if (!retentionMode || !retentionDate) {
-            return false;
-        }
-        if (retentionMode === 'GOVERNANCE' &&
-        headers['x-amz-bypass-governance-retention']) {
-            return false;
-        }
-        const objectDate = moment(retentionDate);
-        const now = moment();
-        // indicates retain until date has expired
-        if (now.isSameOrAfter(objectDate)) {
+
+        if (!this.mode || !this.date) {
             return false;
         }
-        return true;
-    }
-    return false;
-}
-
 
-/* objectLockRequiresBypass will return true if the retention info change
- * would require a bypass governance flag to be true.
- * In order for this to be true the action must be valid as well, so going from
- * COMPLIANCE to GOVERNANCE would return false unless it expired.
- */
-function objectLockRequiresBypass(objectMD, retentionInfo) {
-    const { retentionMode: existingMode, retentionDate: existingDateISO } = objectMD;
-    if (!existingMode) {
-        return false;
+        return !this.isExpired();
     }
 
-    const existingDate = new Date(existingDateISO);
-    const isExpired = existingDate < Date.now();
-    const isExtended = new Date(retentionInfo.date) > existingDate;
+    /**
+     * ObjectLockInfo.isGovernanceMode
+     * @returns {bool} - true if retention mode is GOVERNANCE
+     */
+    isGovernanceMode() {
+        return this.mode === 'GOVERNANCE';
+    }
 
-    if (existingMode === 'GOVERNANCE' && !isExpired) {
-        if (retentionInfo.mode === 'GOVERNANCE' && isExtended) {
-            return false;
-        }
-        return true;
+    /**
+     * ObjectLockInfo.isComplianceMode
+     * @returns {bool} - True if retention mode is COMPLIANCE
+     */
+    isComplianceMode() {
+        return this.mode === 'COMPLIANCE';
     }
 
-    // an invalid retention change or unrelated to bypass
-    return false;
-}
+    /**
+     * ObjectLockInfo.isExpired
+     * @returns {bool} - True if the retention policy policy has expired
+     */
+    isExpired() {
+        const now = moment();
+        return this.date === null || now.isSameOrAfter(this.date);
+    }
 
-function validateObjectLockUpdate(objectMD, retentionInfo, bypassGovernance) {
-    const { retentionMode: existingMode, retentionDate: existingDateISO } = objectMD;
-    if (!existingMode) {
-        return null;
+    /**
+     * ObjectLockInfo.isExtended
+     * @param {string} timestamp - Timestamp in ISP-8601 format
+     * @returns {bool} - True if the given timestamp is after the policy expiration date or if no expiration date is set
+     */
+    isExtended(timestamp) {
+        return (this.date === null && timestamp !== undefined)
+            || (timestamp !== undefined && moment(timestamp).isAfter(this.date));
```

Comment:
small factorization possible

```suggestion
        return timestamp !== undefined &&
          (this.date === null || moment(timestamp).isAfter(this.date));
```